src/components/list/list-item.types.ts

```diff
@@ -40,8 +40,25 @@ export interface ListItem<T = any> {
      * List of actions to display as a menu at the end of the item
      */
     actions?: Array<MenuItem | ListSeparator>;
+
+    /**
+     *Component used to render in the list
+     */
+    component?: ListComponent;
```

Comment:
Since this doesn't replace the whole content of the list-item, I think we should name it something more specific. I wouldn't be surprised if we want to add more slots for custom components in the future, and possibly even an option to completely replace the whole content of the list-item with a custom component. If either of those happen, having this property called `component` is going to be problematic, and renaming it would be a very annoying breaking change 😕

Reply:
My thought was that this was supposed to be used instead of a regular icon and not in addition to it. If it is instead of an icon we could also change `icon` to `string | ListComponent`

Reply:
Thank you @adrianschmidt , if I understood you correctly,  It's should not be called a "component", could it be better if we changed it to "customComponent".  Since we don't know what exactly is used here in the future, right, so quite difficult to be more specific? Or what is your suggestion?

Reply:
good points!

🤔 I don't see a reason for the custom component not to exist with other features like the icons. It's my bad for not providing more use cases. Sorry.…

The intention for having a custom component is that we want to enrich the informativeness and interactivity of the list items. Previously we have managed to hack together lists that are supposed to be informative and interactive. For example the Work-order widget by @eketorp 's team 👇 
![image](https://user-images.githubusercontent.com/35954987/146360585-473f3ff6-bc23-49a5-a610-4a5242cbc344.png)
Even though there is a creative usage of colors and icons to reflect the **status** of an order, it's not always easy for the user to understand that meaning. You only go that far with available tools. But if they had a custom component there, that component could for instance show a string of text, in addition to colors. It could even be interactive and let users take the oder for instance by clicking on it. But it doesn't necessarily have to mean that the icon should be gone. It's then up to the consumer how they want to setup the list. They may just keep the icons, but add other things in the component, or move the icons in there.


Let me share some of our future dreams that we should take into account now.

One scenario is that the custom component should be optimally placed on the right side of the text. The **Update** button in front of **My pipe** in this example is one of them: 👇 
![image](https://user-images.githubusercontent.com/35954987/146360285-6dfe4318-042d-46aa-bf51-6e9122da4c04.png)

At the moment I'm not sure if we should let consumers have `leftCustomComponent` and `rightCostomComponent`, or whether we should let them choose to display their custom component on the right side or left side.

In some scenarios user may want to have two, one on left, and one on right. (Or is this one of those cases that Adrian says, that consumer needs to replace the whole content with their own custom component?)
![145998368-d01ed796-fd0e-408e-8b86-0a5fdd8deb74](https://user-images.githubusercontent.com/35954987/146362024-b0791d38-0bfc-4edf-9c9f-ae2aa50ffcf8.png)

Reply:
> Thank you @adrianschmidt , if I understood you correctly, It's should not be called a "component", could it be better if we changed it to "customComponent". Since we don't know what exactly is used here in the future, right, so quite difficult to be more specific? Or what is your suggestion?

No, it's not that you need to add "custom", it's that the name is very generic, but the placement of the component is quite specific.

Imagine that we would like to add another slot on the right side of the listItem, I can't really come up with anything right now, but let's imagine that we didn't already have the option to render an action menu there, and instead of the solution we have now, where you supply an array of actions, imagine that we decided that we'd have a slot there, where the consumer can add a menu-component of their own.

Then we'd perhaps call that property menuComponent. But, now we have one property menuComponent, and one property component. That's a little odd, right?

Then perhaps we want to make it possible for the consumer to supply a component that replaces the whole standard list-item. What should we call that? We might call it replacementComponent perhaps, but now we have menuComponent for adding a menu, replacementComponent for replacing the whole list-item, and component for adding a little thing on the left side? That would make no sense :sweat_smile:

So we need to come up with a good name for that "little thing on the left side" :smile: And preferably something that doesn't include "left" :smile:

Reply:
@Kiarokh Do you have any suggestions?

If we don't come up with anything better, I guess we could call it something like `widget`. Then _at least_ we keep `component` free for if we want to let the consumer replace the whole item one day 😅

Reply:
`iconAdjacentWidget` 😂 🙈

Reply:
> One scenario is that the custom component should be optimally placed on the right side of the text. The **Update** button in front of **My pipe** in this example is one of them: 👇 ![image](https://user-images.githubusercontent.com/35954987/146360285-6dfe4318-042d-46aa-bf51-6e9122da4c04.png)

This feels like "action" to me.

> At the moment I'm not sure if we should let consumers have `leftCustomComponent` and `rightCostomComponent`, or whether we should let them choose to display their custom component on the right side or left side.
>
> In some scenarios user may want to have two, one on left, and one on right. (Or is this one of those cases that Adrian says, that consumer needs to replace the whole content with their own custom component?) ![145998368-d01ed796-fd0e-408e-8b86-0a5fdd8deb74](https://user-images.githubusercontent.com/35954987/146362024-b0791d38-0bfc-4edf-9c9f-ae2aa50ffcf8.png)

This one could be "primary" and "secondary", respectively.

> Another example: <img alt="Screenshot 2021-12-16 at 12 24 30" width="1021" src="https://user-images.githubusercontent.com/35954987/146363187-15121747-2a01-4b69-8614-4f8a069af64b.png">

This one could perhaps also be "secondary".

So, we could use `primaryComponent` perhaps, leaving the door open for `secondaryComponent` to be placed on the right side, and, if we feel the need to differentiate between "secondary" and "action", we could add `actionComponent` one day.

Unless someone has a better idea 😊

Reply:
Or should we call them `leftWidget`, and `rightWidget`?

If someone decides to use the `leftWidget` or `rightWidget` as a 100%wide widget, then they can be my guest :D

Reply:
I think `left` and `right` are less than optimal. RTL languages, we might want to move them (`.greenHeading { color: blue; }`), stuff like that…

Reply:
I really don't like the term `widget`, it feels like we are introducing another concept very similar to something we already have

Reply:
Right, I see your point. Nowadays, in CSS the use `block-start` and `block-end` to be compatible with different writing directions. For example instead of saying `margin-right`, they write `margin-block-end`. [Read more…](https://developer.mozilla.org/en-US/docs/Web/CSS/margin-block-end) Not sure if it's a good idea to borrow this naming convention somehow?

Reply:
primary and secondary would of course work as well